### ISSUE TITLE
Recursive CTE not working as expected

### ISSUE THREAD

**Issue by alvinr** [tempesta-tech/mariadb#174]
# Problem
In a prior release (may be 0.7) recursive CTE expressed worked as expected. However with a recent commit only a single level appears to be recursed with as 'as of' query.

```sql
with recursive
ancestors
as
(
  select e.emp_id, e.name, e.mgr, e.salary
  from emp as e
  where name = 'bill'
  union 
  select e.emp_id, e.name, e.mgr, e.salary
  from emp as e, ancestors as a
  where e.mgr = a.emp_id
)
select * from ancestors;
+--------+------+------+--------+
| emp_id | name | mgr  | salary |
+--------+------+------+--------+
|      1 | bill | NULL |   2000 |
|     30 | jane |    1 |    750 |
|     20 | john |   30 |    500 |
+--------+------+------+--------+
```

```sql
 with recursive
ancestors
 as
(
  select e.emp_id, e.name, e.mgr, e.salary
  from emp as e for system_time as of timestamp @ts_2
  where name = 'bill'
  union 
  select e.emp_id, e.name, e.mgr, e.salary
  from emp as e for system_time as of timestamp @ts_2, 
       ancestors as a
  where e.mgr = a.emp_id
)
select * from ancestors;
+--------+------+------+--------+
| emp_id | name | mgr  | salary |
+--------+------+------+--------+
|      1 | bill | NULL |   2000 |
|     30 | jane |    1 |    750 |
+--------+------+------+--------+
```

Git Commit: fa51a8270b2c791d85cebfb749238487c645fb3e

# Reproduce
```sql
create or replace table dept (
  dept_id int(10) primary key, 
  name varchar(100)
)
with system versioning
;

create or replace table emp (
  emp_id int(10) primary key,
  dept_id int(10) not null,
  name varchar(100) not null,
  mgr int(10),
  salary int(10) not null,
  constraint `dept-emp-fk`
    foreign key (dept_id) references dept (dept_id)
      on delete cascade
      on update restrict,
  constraint `mgr-fk`
    foreign key (mgr) references emp (emp_id)
      on delete restrict
      on update restrict
) 
with system versioning;

insert into dept (dept_id, name) values (10, "accounting");

insert into emp (emp_id, name, salary, dept_id, mgr) values 
(1, "bill", 1000, 10, null),
(20, "john", 500, 10, 1),
(30, "jane", 750, 10,1 );

select vtq_commit_ts(max(sys_trx_start)) into @ts_1 from emp;

update emp set mgr=30 where name ="john";
select vtq_commit_ts(sys_trx_start) into @ts_2 from emp where name="john";

/* All report to 'Bill' */
with recursive
ancestors
as
(
  select e.emp_id, e.name, e.mgr, e.salary
  from emp as e for system_time as of timestamp @ts_1
  where name = 'bill'
  union 
  select e.emp_id, e.name, e.mgr, e.salary
  from emp as e for system_time as of timestamp @ts_1, 
       ancestors as a
  where e.mgr = a.emp_id
)
select * from ancestors;

/* Only two rows shown, not three */
with recursive
ancestors
as
(
  select e.emp_id, e.name, e.mgr, e.salary
  from emp as e for system_time as of timestamp @ts_2
  where name = 'bill'
  union 
  select e.emp_id, e.name, e.mgr, e.salary
  from emp as e for system_time as of timestamp @ts_2,
       ancestors as a
  where e.mgr = a.emp_id
)
select * from ancestors;
```

# Workaround
None

# Solution
The 'as of' version of the CTE should traverse the full hierarchy.

**Comment by kevgs** [comment]
```sql
with recursive
ancestors
as
(
  select e.emp_id, e.name, e.mgr, e.salary
  from emp as e for system_time as of timestamp @ts_2
  where name = 'bill'
  union 
  select e.emp_id, e.name, e.mgr, e.salary
  from emp as e for system_time as of timestamp @ts_2,
       ancestors as a
  where e.mgr = a.emp_id
)
select * from ancestors;
```

Query consists of inner and outer parts. Inner part returns 2 alive and 1 historical row. But, outer part uses default `for system_time` restriction (`sys_trx_end` == `max_trx_id`) and finally allows just 2 alive rows. To see all 3 rows you need to write
```sql
...
select * from ancestors for system_time as of timestamp @ts_2;
```

This is how it works now. Outer query and views see only what its inner part returns. Several nested levels are allowed but historical rows do not become alive ones and require `for system_time ...` to be shown.

Please, check your test code again on a code after this commit.
https://github.com/tempesta-tech/mariadb_10.2/commit/7cad7c637638f3d179692c354d9d027e7a931b54

Looking forward for your feedback.

**Comment by midenok** [comment]
Related to #74, #98, #183.

**Comment by midenok** [comment]
Fixed by #175.
This works now, all 3 rows returned:
```
+--------+------+------+--------+
| emp_id | name | mgr  | salary |
+--------+------+------+--------+
|      1 | bill | NULL |   1000 |
|     30 | jane |    1 |    750 |
|     20 | john |   30 |    500 |
+--------+------+------+--------+
```
